### DISCUSSION TITLE
Ping request not sent under heavy load?

### DISCUSSION THREAD

**Discussion by jhoughjr** [adam-fowler/mqtt-nio#43] [Q&A]
Moving the discussion here.

A summary;
My setup is Mosquitto on macOS as the broker.
iOS client using MQTTNIO and a C++ client running on wio terminal.

The WIO publishes 3 topic of 1-3 bytes each set to every ten µs.
From the Mosquitto logs that rate does not appear to be making it through.

After about 500 messages on each topic, the iOS client disconnects.
searches for PINGREQ show only the wio terminal client.

I'll XP by reducing the interval the WIO sends at.

**Comment by jhoughjr** [comment]
Im also looking for a breakpoint when the ping req is sent as I seem to have found the code.

**Reply by jhoughjr** [reply]
else block 68-70 of PingreqHandler was reached . the condition self.lastEventTime + self.timeout <= .now() was false

**Comment by adam-fowler** [comment]
Jimmy is this with iOS 15 beta
I've been running the subscribe on the iOS 14 emulator for the last 10 minutes
I'll setup an iOS project and run it on my phone

**Reply by jhoughjr** [reply]
yeah its iOS 15

some logging im doing to see what is causing it:

It hit this else block before the disconnect timeout it seems.

(lldb) po self.lastEventTime
▿ 58990995073666
  - _uptimeNanoseconds : 58990995073666

▿ 58990995073666
  - _uptimeNanoseconds : 58990995073666

(lldb) po self.timeout
▿ TimeAmount
  - nanoseconds : 85000000000

(lldb) po NIODeadline.now()
▿ 59042919584625
  - _uptimeNanoseconds : 59042919584625

(lldb)

**Reply by adam-fowler** [reply]
Can you reduce the pingInterval to 5 seconds and see if it ever gets sent

**Reply by jhoughjr** [reply]
yeah ill check that I had started to use the big initer since I cant set it on the conf afterwards, but then decided to dig into the code.

**Reply by jhoughjr** [reply]
Xcode is not finding the type of TimeAmount :/

**Reply by adam-fowler** [reply]
Its in NIO

**Reply by jhoughjr** [reply]
using this config, it crashes

config
```
Optional(MQTTNIO.MQTTClient.Configuration(disablePing: false, keepAliveInterval: NIO.TimeAmount(nanoseconds: 30000000000), pingInterval: Optional(NIO.TimeAmount(nanoseconds: 5000000000)), timeout: Optional(NIO.TimeAmount(nanoseconds: 30000000000)), maxRetryAttempts: 3, userName: nil, password: nil, useSSL: false, useWebSockets: false, tlsConfiguration: nil, sniServerName: nil, webSocketURLPath: nil))```

```
2021-06-23 10:04:34.055639-0500 MQTTTesting[2413:697827] -[__NSCFNumber countByEnumeratingWithState:objects:count:]: unrecognized selector sent to instance 0x8000000000000000
2021-06-23 10:04:34.061630-0500 MQTTTesting[2413:697827] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFNumber countByEnumeratingWithState:objects:count:]: unrecognized selector sent to instance 0x8000000000000000'
*** First throw call stack:
(0x18403ae00 0x19b04e0d8 0x1840c5f70 0x183f7e1ec 0x183fdbbfc 0x188645d60 0x18856be84 0x18856c0b4 0x18858b324 0x102b913d0 0x102b7e730 0x102b91648 0x102d1e9b4 0x102b91148 0x102b7b99c 0x102b7a7e4 0x102b7d760 0x102c34d04 0x102c3b4d0 0x102c33c7c 0x102c34644 0x102c35b28 0x102c361e8 0x102c1a770 0x102c1a7c8 0x102c1e210 0x102b9d1dc 0x102b9d71c 0x102c1a770 0x102c17484 0x102c17370 0x102d94270 0x102d93e5c 0x1dc391824 0x1dc3914e0 0x1038c10c0 0x1038b17c4 0x1038b8d44 0x1038b9a48 0x1038b8bd0 0x1038b9a48 0x1038c5864 0x1dc6165c0 0x1dc619a40)
libc++abi: terminating with uncaught exception of type NSException
dyld4 config: DYLD_LIBRARY_PATH=/usr/lib/system/introspection DYLD_INSERT_LIBRARIES=/Developer/usr/lib/libBacktraceRecording.dylib:/Developer/usr/lib/libMainThreadChecker.dylib:/Developer/Library/PrivateFrameworks/DTDDISupport.framework/libViewDebuggerSupport.dylib
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFNumber countByEnumeratingWithState:objects:count:]: unrecognized selector sent to instance 0x8000000000000000'
terminating with uncaught exception of type NSException
(lldb) ```

**Reply by jhoughjr** [reply]
the crash might have been due to a breakpoint, Nother thing to note, when it disconnected cancelTask was reached as I still had a bp there.

**Reply by jhoughjr** [reply]
still no ping messages logged in Mosquitto. Im going to try and slow the wio publish down. (been trying for realtime visualization :) )

**Reply by jhoughjr** [reply]
I still dont think the ping is even being sent rather than it failing. Still digging. Ill make a repo for this so you can see the code. Something might pop out.

**Reply by adam-fowler** [reply]
Looking through the code it only will ever send a ping if the connection is inactive. I have seen a possible bug though. I reset the timer on both read and write of messages when this should only ever be on write.

Are your messages being sent with QoS 0 (.atMostOnce)?

**Reply by adam-fowler** [reply]
If this is the case, you should try the `1.x.x` branch which has a fix for this

**Reply by adam-fowler** [reply]
@jhoughjr did the `1.x.x` branch fix your issue
